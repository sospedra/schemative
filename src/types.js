import { assign, reduce, noop } from 'lodash'
import PropTypes from 'prop-types'

const EXPOSED_PROPERTIES = ['value']

/**
 * Trigger a warning when assigning a value which does not
 * fit with the provided scalar.
 * Return default value.
 *
 * @param  {Any} candidate
 * @param  {Any} fallback
 * @param  {String} scalar
 * @return {Any}
 */
const evaluateValueType = (candidate, fallback, scalar) => {
  if (scalar !== 'any' && typeof candidate !== typeof fallback) {
    console.warn(
      'Evalute Schemative type expected',
      typeof fallback, 'but receive', typeof candidate,
      'when', candidate, 'value was set.',
      'Default value is returned instead.'
    )

    return fallback
  }

  return candidate
}

/**
 * Given a value, type base and default value creates an Schemative type.
 * Create the object with the type definition.
 * Pick all the properties that goes to the prototype.
 * Add the properties to an object containing the type exposed values.
 *
 * @param  {Any} value        - Selected value
 * @param  {Object} base      - Contains basic type's data
 *   @param  {String} scalar    - Type name
 *   @param  {Func} propTypes   - React PropTypes equivalent
 * @param  {Any} defaultValue - Default type value
 * @return {Object}           - Type
 */
const createType = (value, base, defaultValue) => {
  const type = assign({}, base, {
    isRequired: !base.__required__ &&
      createType(value, createTypeBase(base, true), defaultValue)
  })
  const properties = reduce(type, (memo, value, property) => {
    if (EXPOSED_PROPERTIES.includes(property)) return memo

    return assign({}, memo, {
      [property]: { get: () => value }
    })
  }, {})

  return Object.defineProperties({
    value: evaluateValueType(value, defaultValue, base.scalar)
  }, properties)
}

const createTypeBase = (base, isRequired) => {
  const propType = PropTypes[base.scalar]

  return {
    ...base,
    __required__: !!isRequired,
    propType: isRequired ? propType.isRequired : propType,
    scalar: base.scalar
  }
}

const createTypeWithDefault = (scalar, defaultValue) => {
  const base = createType(defaultValue, createTypeBase({ scalar }), defaultValue)
  const type = (value) => createType(value, base, defaultValue)

  Object.getOwnPropertyNames(base).forEach((property) => {
    if (EXPOSED_PROPERTIES.includes(property)) {
      type[property] = base[property]
    } else {
      Object.defineProperty(type, property, {
        get: () => base[property]
      })
    }
  })

  return type
}

export const any = createTypeWithDefault('any', true)
export const array = createTypeWithDefault('array', [])
export const bool = createTypeWithDefault('bool', true)
export const element = createTypeWithDefault('element', noop)
export const func = createTypeWithDefault('func', noop)
export const number = createTypeWithDefault('number', 0)
export const object = createTypeWithDefault('object', {})
export const shape = createTypeWithDefault('shape', {})
export const string = createTypeWithDefault('string', '')
export const symbol = createTypeWithDefault('symbol', Symbol(''))
