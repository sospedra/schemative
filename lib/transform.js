'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.transform = exports.createTransform = exports.filterByKeys = undefined;

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var protoMutators = {};
var defsKeys = {};

var filterByKeys = exports.filterByKeys = function filterByKeys(defsKeys, candidate) {
  var keys = _lodash2.default.intersection(defsKeys, _lodash2.default.keys(candidate));
  return _lodash2.default.pick(candidate, keys);
};

var createTransform = exports.createTransform = function createTransform(definitions) {
  transform.keys = _lodash2.default.keys(definitions);
  return transform;
};

var transform = exports.transform = function transform(candidate) {
  var mutators = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : protoMutators;

  return Object.assign(filterByKeys(defsKeys, candidate), _lodash2.default.reduce(mutators, function (memo, value, key) {
    return Object.assign(_defineProperty({}, key, _lodash2.default.isFunction(value) ? value(candidate) : value), memo);
  }, {}));
};

Object.defineProperty(transform, 'mutators', {
  get: function get() {
    return protoMutators;
  },
  set: function set(mutators) {
    protoMutators = mutators;
  }
});

Object.defineProperty(transform, 'keys', {
  get: function get() {
    return defsKeys;
  },
  set: function set(definitions) {
    defsKeys = definitions;
  }
});
//# sourceMappingURL=transform.js.map