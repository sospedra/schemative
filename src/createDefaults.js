import tsil from 'tsil'

export default (nodes) => {
  const defaultValues = tsil.modify(nodes, (node) => {
    return node.value
  })

  return tsil.deflatten(defaultValues)
}
