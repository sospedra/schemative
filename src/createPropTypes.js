import tsil from 'tsil'

export default (nodes) => {
  const propTypes = tsil.modify(nodes, (node) => {
    return node.propType
  })

  return tsil.deflatten(propTypes)
}
