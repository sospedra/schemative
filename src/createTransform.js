export default (nodes) => {
  // console.log('createTransform')
}
