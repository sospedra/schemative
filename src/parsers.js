import _ from 'lodash'

export const getPropType = (prop, attr) => {
  return prop.iterable ? recursiveValues(prop, attr) : prop[attr]
}

export const recursiveObject = (values, attr) => {
  return _.reduce(values, (memo, prop, key) => Object.assign({
    [key]: getPropType(prop, attr)
  }, memo), {})
}

export const recursiveArray = (values, attr) => {
  return _.map(values, (prop) => {
    return getPropType(prop, attr)
  })
}

export const selectRecursiveStrategy = (values, attr) => {
  return _.isArray(values)
    ? recursiveArray(values, attr)
    : recursiveObject(values, attr)
}

export const recursiveValues = ({ type, values }, attr) => {
  const exec = attr === 'type' ? type : _.identity
  return exec(selectRecursiveStrategy(values, attr))
}

export const createPropTypes = (definition) => {
  return recursiveObject(definition, 'type')
}

export const createDefault = (definition) => {
  return recursiveObject(definition, 'default')
}

export const filterByKeys = (definition, candidate) => {
  const keys = _.intersection(_.keys(definition), _.keys(candidate))
  return _.pick(candidate, keys)
}

export const transform = (definition, candidate, mutators) => Object.assign(
  filterByKeys(definition, candidate),
  _.reduce(mutators, (memo, value, key) => Object.assign({
    [key]: _.isFunction(value) ? value(candidate) : value,
    memo
  }), {})
)
