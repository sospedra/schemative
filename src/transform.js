import _ from 'lodash'

export const filterByKeys = (defsKeys, candidate) => {
  const keys = _.intersection(defsKeys, _.keys(candidate))
  return _.pick(candidate, keys)
}

export const createTransform = (definitions) => {
  const defsKeys = _.keys(definitions)
  let protoMutators = {}

  const transform = (candidate, mutators = protoMutators) => {
    return Object.assign(
      filterByKeys(defsKeys, candidate),
      _.reduce(mutators, (memo, value, key) => Object.assign({
        [key]: _.isFunction(value) ? value(candidate, definitions) : value
      }, memo), {})
    )
  }

  Object.defineProperty(transform, 'mutators', {
    get: () => protoMutators,
    set: (mutators) => {
      protoMutators = mutators
    }
  })

  return transform
}
