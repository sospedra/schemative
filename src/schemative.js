import { PropTypes } from 'react'
import _ from 'lodash/util'
import { createPropTypes, createDefault } from './parsers'
import { createTransform } from './transform'

const boundIsRequire = (prototype) => {
  return Object.defineProperty(prototype, 'isRequired', {
    value: Object.assign({}, prototype, {
      type: prototype.type.isRequired
    })
  })
}

const createType = (type, def) => boundIsRequire({
  type: PropTypes[type],
  default: def,
  iterable: false
})

const createFunctionType = (type) => (values) => boundIsRequire({
  type: PropTypes[type],
  values: values,
  iterable: true
})

export const any = createType('any', _.stubTrue())
export const array = createType('array', _.stubArray())
export const bool = createType('bool', _.stubTrue())
export const element = createType('element', _.stubFalse())
export const func = createType('func', _.noop())
export const instanceOf = createType('instanceOf', _.stubFalse())
export const number = createType('number', 0)
export const node = createType('node', _.stubFalse())
export const object = createType('object', _.stubObject())
export const string = createType('string', _.stubString())
export const symbol = createType('symbol', Symbol)
export const arrayOf = createFunctionType('arrayOf')
export const objectOf = createFunctionType('objectOf')
export const oneOf = createFunctionType('oneOf')
export const oneOfType = createFunctionType('oneOfType')
export const shape = createFunctionType('shape')

export const createSchema = (definition) => ({
  PropTypes: createPropTypes(definition),
  Default: createDefault(definition),
  transform: createTransform(definition)
})
