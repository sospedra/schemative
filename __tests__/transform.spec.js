/* global describe, it, expect */
import { filterByKeys, transform } from '../src/transform'
import * as mocks from './mocks'

describe('Transform suite', () => {
  it('should filter by keys two objects', () => {
    const keys = Object.keys(mocks.defs.simple)
    expect(filterByKeys(keys, mocks.defs.complex)).toEqual(mocks.defs.simple)
  })

  it('should transform a definition into a fulfilled object', () => {
    expect(transform(mocks.defs.simple, mocks.fulfilled)).toEqual(mocks.fulfilled)
  })

  it('should transform a definition using mutators', () => {
    expect(transform(mocks.defs.simple, mocks.fulfilled, mocks.mutators)).toMatchSnapshot()
  })

  it('should be able to define agent mutators', () => {
    transform.mutators = { foo: 'bar' }
    expect(transform(mocks.defs.simple, mocks.fulfilled)).toMatchSnapshot()
  })
})
