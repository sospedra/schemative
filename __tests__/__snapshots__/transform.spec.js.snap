exports[`Transform suite should be able to define agent mutators 1`] = `
Object {
  "foo": "bar",
  "name": "Ada Lovelace",
  "year": 1815,
}
`;

exports[`Transform suite should transform a definition using mutators 1`] = `
Object {
  "extra": "attribute",
  "name": "Ada Lovelace",
  "year": 1852,
}
`;
